### ISSUE TITLE
Intermittent failure in PB tests

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2327]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2327|
|Type|defect|
|Created|2006-12-24 18:44:13Z|
|Branch|https://github.com/twisted/twisted/tree/test_publishable-win32-2327|

```
===============================================================================
[ERROR]: twisted.test.test_pb.BrokerTestCase.testPublishable

Traceback (most recent call last):
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_pb.py", line 651, in testPublishable
    pump.flush()
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_pb.py", line 71, in flush
    while self.pump():
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_pb.py", line 93, in pump
    self.client.dataReceived(byte)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\banana.py", line 218, in dataReceived
    gotItem(item)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\banana.py", line 148, in gotItem
    self.callExpressionReceived(item)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\banana.py", line 113, in callExpressionReceived
    self.expressionReceived(obj)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\pb.py", line 523, in expressionReceived
    method(*sexp[1:])
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\pb.py", line 892, in proto_answer
    d.callback(self.unserialize(netResult))
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\pb.py", line 784, in unserialize
    return unjelly(sexp, self.security, None, self)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\jelly.py", line 900, in unjelly
    return _Unjellier(taster, persistentLoad, invoker).unjellyFull(sexp)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\jelly.py", line 523, in unjellyFull
    o = self.unjelly(obj)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\jelly.py", line 546, in unjelly
    val = method(self, obj)
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\flavors.py", line 458, in unjellyFor
    cProxy.setCopyableState(unjellier.unjelly(jellyList[2]))
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\publish.py", line 86, in setCopyableState
    newself = jelly.unjelly(banana.decode(data))
  File "c:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\spread\banana.py", line 349, in decode
    return l[0]
exceptions.IndexError: list index out of range
-------------------------------------------------------------------------------
```

I expect this is a test bug related to the mock transport implemented and used by these tests, rather than a bug in PB itself.

<details><summary>Searchable metadata</summary>

```
trac-id__2327 2327
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_test_publishable_win32_2327 branches/test_publishable-win32-2327
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1166985853000000 1166985853000000
changetime__1215609129000000 1215609129000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#2879](https://github.com/twisted/twisted/issues/2879) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24177]) Branching to 'test_publishable-win32-2327'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I don't know how this test could have ever passed.  It wrote banana data, including `\x1a` (aka EOF on Windows) to a binary file.  Later, it tried to read it back from a text file, resulting in truncated data.  Should have always failed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

It looks good. It would be nice though if the file were explicitly closed (same thing in _cbGotUpdate). But please merge when you want.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [24241]) Explicitly close opened files

refs [#2327](https://github.com/twisted/twisted/issues/2327)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [24244]) Merge test_publishable-win32-2327

Author: exarkun
Reviewer: therve
Fixes: [#2327](https://github.com/twisted/twisted/issues/2327)

Read publishable data files in binary mode so that the data is not corrupted on
Windows.  Also, explicitly close the files rather than relying on refcounting to
do it.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
